traits/trait_list_object.py

```diff
@@ -53,460 +88,755 @@ def __init__(self, index=0, removed=None, added=None):
-
-                index = len(self) + key if key < 0 else key
-
-            list.__setitem__(self, key, value)
-            if self.name_items is not None:
-                if delta == 0:
-                    try:
-                        if removed == values:
-                            return
-                    except Exception:
-                        # Treat incomparable values as equal:
-                        pass
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(index, removed, values)
-                )
+    def __getstate__(self):
+        """ Get the state of the object for serialization.
 
-        except TraitError as excp:
-            excp.set_prefix("Each element of the")
-            raise excp
+        Notifiers are transient and should not be serialized.
+        """
+        result = self.__dict__.copy()
+        result.pop("notifiers", None)
+        return result
 
-    def __delitem__(self, key):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            return list.__delitem__(self, key)
+    def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
 
-        try:
-            removed = self[key]
-        except Exception:
-            removed = []
+        Notifiers are transient and are restored to the empty list.
+        """
+        state['notifiers'] = []
+        self.__dict__.update(state)
 
-        if isinstance(key, slice):
-            slice_len = len(removed)
-            delta = slice_len
-            step = 1 if key.step is None else key.step
-            if step == 1:
-                # FIXME: See corresponding comment in __setitem__() for
-                # explanation.
-                index = 0 if key.start is None else key.start
-            else:
-                index = key
-        else:
-            delta = 1
-            index = len(self) + key + 1 if key < 0 else key
-            removed = [removed]
+    def __setitem__(self, index, value):
+        """ Set a value at index, implements self[index] = value
+
+        Parameters
+        ----------
+        index : int
+            Index at which the value will be set.
+        value : Any
+            The value to set at the index.
+
+        Notes
+        -----
+        Notification:
+            index : int
+                The index of the value.
+            added : Any
+                The value that is set.
+            removed : Any
+                The removed value or Undefined if no value was previously
+                present at the index.
 
-        if not (trait.minlen <= (len(self) - delta)):
-            self.len_error(len(self) - delta)
-            return
+        """
+        removed = self._get_removed(index)
 
-        list.__delitem__(self, key)
+        if isinstance(index, slice):
+            if len(removed) != len(value) and index.step not in {1, None}:
+                raise ValueError
```

Comment:
It will, but this check is necessary for backward compatibility. Failing tests otherwise.

Reply:
OK. Then could you instantiate the exception with the error message matching what could otherwise come from Python please?
(If that's ok, please let me resolve my own comment.)

Reply:
> OK. Then could you instantiate the exception with the error message matching what could otherwise come from Python please?

I am sorry that I probably should have included this in the initial comment instead of assuming the branch can be removed.

I confirmed removing this results in a test failure because we get a different exception type. The exception changed from ValueError to TraitError because the validation raises.

Reply:
Getting the right error message involves calculating the size of the slice , it can get a little tricky, do you think its necessary ?

Reply:
I thought the error message is just this?:
```
"attempt to assign sequence of size {} to extended slice of size {}".format(len(value), len(removed))
```

Reply:
Can `removed` cause an `IndexError` for slices ? If not what you suggested will work

Reply:
I guess not.  Ok, updating it.

Reply:
I unresolved this conversation because it looks as though the issue is still valid: we should have at least _some_ kind of message for the `raise ValueError`.